4.1.x

ileinone opened the pull request:
Adding functionality like audit trail, logging, cryptography, and more fine grained exception handling among other features executed as aspects.

snicoll committed:
Next Development Version

jhoeller committed:
Polishing
(cherry picked from commit 730bd02)

jhoeller committed:
Polishing

jhoeller committed:
Polishing

bclozel committed:
Revisit empty body response support in HTTP client

Prior to this commit, HTTP responses without body (response status 204
or 304, Content-Length: 0) were handled properly by RestTemplates. But
some other cases were not properly managed, throwing exceptions for
valid HTTP responses.

This commit better handles HTTP responses, using a response wrapper that
can tell if a response:

* has no message body (HTTP status 1XX, 204, 304 or Content-Length:0)
* has an empty message body

This covers rfc7230 Section 3.3.3.

Issue: SPR-8016

snicoll committed:
Restore default transaction manager by name lookup

Fix a regression introduced by 961574bd17b that prevents a proper lookup
of the default transaction manager by name as the absence of a qualifier
is represented by an empty string (passing the faulty null check).

Issue: SPR-12577

jhoeller committed:
ObjectToOptionalConverter uses Optional.ofNullable after ConversionService invocation

Issue: SPR-12589
(cherry picked from commit ec84fa6)

jhoeller committed:
Revised ExtendedBeanInfo test for SPR-8937 (for JDK 8u40 compatibility)

Issue: SPR-12582
(cherry picked from commit 7492129)

sbrannen committed:
Handle exceptions properly in SpringJUnit4ClassRunner

JUnit 4.9 introduced a regression in BlockJUnit4ClassRunner.runChild()
such that exceptions thrown from methodBlock() cause the current test
execution to abort immediately. As a result, the failing test method is
unrooted, and subsequent test methods are never invoked. Furthermore,
RunListeners registered with JUnit are not properly notified.

In conjunction with SPR-11908, SpringJUnit4ClassRunner was updated to
use the aforementioned changes to BlockJUnit4ClassRunner.runChild().
Consequently, SpringJUnit4ClassRunner now suffers from the same
regression.

This commit addresses this issue by ensuring that any exceptions thrown
during the invocation of methodBlock() are properly wrapped in a JUnit
Fail Statement.

Issue: SPR-12613
(cherry picked from commit b81c522ee1624f3d5f170fc3468f4600956950bd)

sbrannen committed:
Enable reuse of DefaultActiveProfilesResolver

In order to allow DefaultActiveProfilesResolver to be reused (e.g., via
extension or delegation), the check which asserts that the 'resolver'
attribute of @ActiveProfiles is not set to a customer resolver class
has been removed.

Issue: SPR-12611
(cherry picked from commit 276712dcd1eef93e2a9add83e3d11f0eb6f06095)

bclozel committed:
Fix context-relative default value in XML parsing

This commit fixes the default value for the contextRelative attribute of
a RedirectView, when this view is registered via a
RedirectViewController in XML. The value is set to true.

Note that the default value for this is correctly documented in
spring-mvc-4.1.xsd. Also, the documentation and implementation for its
javadoc counterpart also enforces true as a default value.

Issue: SPR-12607

Commit:
Fix method documentation typo

Issue: SPR-12639

bclozel committed:
Allow protocol relative URLs in CssLink Transformer

This commit allows the use of "protcol relative URLs" (i.e. URLs without
scheme, starting with `//`), often used to serve resources automatically
from https or http with third party domains.

This syntax is allowed by RFC 3986.

Issue: SPR-12632

jhoeller committed:
AnnotationUtils explicitly handles null parameters

Issue: SPR-12604
(cherry picked from commit 0ddf8dd)

jhoeller committed:
AbstractRequestLoggingFilter allows for dedicated shouldLog check in CommonsRequestLoggingFilter

Issue: SPR-12609
(cherry picked from commit d4dac25)

jhoeller committed:
AbstractBeanDefinitionParser allows for skipping evaluation of XML "name" attribute

Issue: SPR-12643
(cherry picked from commit 11bf3b3)

jhoeller committed:
OperatorMatches caches compiled patterns

Issue: SPR-12610
(cherry picked from commit d34402d)

jhoeller committed:
Latest dependency updates (Jackson 2.5, Hibernate 4.3.8, Netty 4.0.25)
(cherry picked from commit 4141bf3)

bclozel committed:
Allow relative paths within resource location path

Prior to this change, location paths used for resource handling would
not allow "non-cleaned, relative paths" such as
`file://home/user/static/../static/`. When checking if the resolved
resource's path starts with the location path, a mismatch would happen
when comparing for example:

* the location `file://home/user/static/../static/`
* and the resource `file://home/user/static/resource.txt`

This commit cleans the location path before comparing it to the resource
path.

Issue: SPR-12624

rstoyanchev committed:
Fix minor test issues

Issue: SPR-12615

rstoyanchev committed:
Remove logging statement in ResponseBodyAdviceChain

Issue: SPR-12616

bclozel committed:
Initialize ResourceUrlProvider only once

Prior to this change, the ResourceUrlProvider would listen to
ContextRefreshedEvents and autodetect resource handlers each time. This
can cause issues when multiple contexts are involved and the last one
has no resource handler, thus clearing the previously detected ones.

This commit disables resource handlers auto-detection once some have
been detected with a refreshed context.

Issue: SPR-12592

jhoeller committed:
Allow schedulerWithHsqlDataSource to pass through reducing it to the trigger table check

Issue: SPR-12618
(cherry picked from commit 49e31c3)

jhoeller committed:
ScheduledAnnotationBeanPostProcessor registers tasks in ContextRefreshedEvent phase (again)

Issue: SPR-12641
(cherry picked from commit 0479ca6)

jhoeller committed:
Polishing
(cherry picked from commit 1cd4433)

jhoeller committed:
Latest applicable dependency updates (Jackson 2.4.5, Jetty 9.2.7)

rstoyanchev committed:
Remove ISE in ResourceUrlProvider

Issue: SPR-12630

rstoyanchev committed:
Add JdkIdGenerator and use it in SockJS client

Issue: SPR-12658

bclozel committed:
Avoid loss of body content in AbstractRequestLoggingFilter

Prior to this commit, the `ContentCachingRequestWrapper` class would
cache the response content only if the reponse would be consumed using
its InputStream. In case of a Form request, Spring MVC consumes the
response using the `getParameter*` Servlet API methods. This causes the
cached content to never be written.

This commit makes the `ContentCachingResponseWrapper` write the request
body to the cache buffer by using the `getParameter*` API, thus avoiding
those issues.

Issue: SPR-7913

sbrannen committed:
Support @Configuration as meta-annotation in the TCF

Spring Framework 4.0 introduced support for using test-related
annotations as meta-annotations in the Spring TestContext Framework
(TCF) in order to create custom composed annotations within a test
suite; however, the detection of default @Configuration classes in test
classes was not updated to search for @Configuration declared as a
meta-annotation. Specifically, AnnotationConfigContextLoaderUtils
invokes Class.isAnnotated() which only searches for annotations
declared directly on the class in question.

This commit addresses this issue by refactoring the
isDefaultConfigurationClassCandidate() method in
AnnotationConfigContextLoaderUtils so that it uses
AnnotationUtils.findAnnotation() instead of Class.isAnnotated() for
detecting the presence of the @Configuration annotation, either
directly or as a meta-annotation.

Issue: SPR-12659
(cherry picked from commit 2d918380f0553238087ec264688ecba066e49ab5)

snicoll committed:
Allow subclasses to configure the Yaml instance

Provide an additional hook-point for YamlProcessor subclasses willing to
change how the Yaml instance is configured. Also expose the default
StrictMapAppenderConstructor so that  they can compose a custom instance
with it.

Issue: SPR-12671

rstoyanchev committed:
DefaultSubscriptionRegistry returns safe to iterate Map

Prior to this change when adding subscriptions
DefaultSubscriptionRegistry (incorrectly) made a copy of the given map
for its "access" cache rather than for its "update" cache.

Issue: SPR-12665

rstoyanchev committed:
Update Netty4ClientHttpRequestFactory buffer size

This change deprecates the maxRequestSize property and adds
maxResponseSize instead. The latter is required to create Netty's
HttpObjectAggregator and aggregates responses.

The maxRequestSize property is already removed in the master branch
and will not be available starting with 4.2.

Issue: SPR-12623

rstoyanchev committed:
Update javadoc for MultipartFile.transferTo

Issue: SPR-12650

bclozel committed:
Sort handler matches in ResourceUrlProvider

Prior to this change, the `ResourceUrlProvider.getForLookupPath` method
would try to match handlers using the keySet order in the
handlerMappings Map. In case of several matches, the handler used for
the return value could vary, since the registration order in the
handlerMappings can't be guaranteed in the configuration.

This commit now collects all matching handlers and sort them using a
`PatternComparator`, in order to try each handler from the most specific
mapping to the least.

Issue: SPR-12647

snicoll committed:
Allow MBeans to be excluded additively

Previously, one could only set the list of bean names to exclude from
auto-detection and there was no way to add additional bean names.

MBeanExporter now exposes a addExcludedBean method that can be invoked
during the initialization phase to add bean names to ignore.

Issue: SPR-12686

snicoll committed:
Clarify the use of @Cacheable in PostConstruct code

Update documentation to explicitly mention that the cache interceptor
must be fully initialized to provide the expected behavior and therefore
initialization code should not rely on this feature, i;e. typically in
PostConstruct callback.

Since the Transactional infrastructure has the exact same infrastructure,
update that section of the doc as well.

Issue: SPR-12700

snicoll committed:
Fix broken build

Add required  `@Deprecated` annotation.

sdeleuze committed:
Fix SockJS origin check

This commit introduces the following changes:
 - Requests without Origin header are not rejected anymore
 - Disable Iframe when allowedOrigins is not empty and not equals to *
 - The Iframe is not cached anymore in order to have a reliable origin check
 - allowedOrigins must not be null or empty
 - allowedOrigins format is now validated (should be * or start by http(s)://)

Issue: SPR-12660
(cherry picked from commit 9b3319b)

rstoyanchev committed:
Add fromHttpRequest to UriComponentsBuilder

Before this change, detection of X-Forwarded-* headers was only built
into ServletUriComponentsBuilder.

This change adds a new method for creating a UriComponentsBuilder from
an existing HttpRequest. This is equivalent to the fromUri method +
X-Forwarded-* header values.

jhoeller committed:
Avoid ConcurrentModificationException in getBeansWithAnnotation

Issue: SPR-12688
(cherry picked from commit 918bc3b)

jhoeller committed:
Properly honor conditions on overriding bean methods

Issue: SPR-12694
(cherry picked from commit 981aefc)

jhoeller committed:
Revised validation javadoc, plus protected validate/isBindingErrorFatal template methods

Issue: SPR-12655
(cherry picked from commit 7585be8)

jhoeller committed:
Polishing
(cherry picked from commit 058714b)

jhoeller committed:
Latest dependency updates (JasperReports 6.0.3, Tomcat 8.0.18, Undertow 1.1.2)

snicoll committed:
BeanWrapper auto-grows arrays if necessary

Previously, only indexed access for collections were supported. When
attempting to access the element of an array that had not the requested
size, the call would fail with an IndexOutOfBoundException

This commit harmonize the binding support so that the array is updated
according to the requested index if necessary.

Issue: SPR-12706
(cherry picked from commit aa21339)

wilkinsona committed:
Fix Animal Sniffer

Jasper Reports’ transitive dependency on spring-context (via
castor-xml which is a new dependency in 6.0.3) was being mapped by
Gradle to a dependency on the spring-context project. For reasons that
I do not fully understand this was causing -source and -javadoc jars
to be added to the project's compile classpath which is used by the
Animal Sniffer Ant task. When the task runs these jars do not exist
which causes it to fail. This commit fixes the problem by adding an
exclusion of org.springframework:spring-context to the Jasper Reports
dependencies in spring-context-support and spring-webmvc.
(cherry picked from commit 7a6a132)

jhoeller committed:
Combined backport of Jackson configuration enhancements: Locale/TimeZone settings, refined module configuration

Issue: SPR-12594
Issue: SPR-12634

jhoeller committed:
Latest dependency updates (POI 3.11, FreeMarker 2.3.21)

sbrannen committed:
Add further regression tests for @TestPropertySource

This commit introduces further regression tests to ensure proper parsing
of inlined properties configured via @TestPropertySource. Specifically,
these additional tests ensure that we do not introduce a bug like the
one raised in Spring Boot issue #1110 [0].

[0] https://github.com/spring-projects/spring-boot/issues/1110

Issue: SPR-12710
(cherry picked from commit 67934a22e2f77449dc1fd8e614028e3d42e20e81)

sdeleuze committed:
Change SockJS and Websocket default allowedOrigins to same origin

This commit adds support for a same origin check that compares
Origin header to Host header. It also changes the default setting
from all origins allowed to only same origin allowed.

Issues: SPR-12697, SPR-12685
(cherry picked from commit 6062e15)

rstoyanchev committed:
Pass SockJS session attributes to HandshakeHandler

Before this change the WebSocketTransportHandler passed
Collections.emptyMap as attributes to the HandshakeHandler because
it didn't matter what attributes the underlying WebSocketSession has
since it is wrapped by the SockJsSession and that's what exposed for
use everywhere.

This change has the WebSocketTransportHandler passing the attributes
from the SockJsSession instead since it's more accurate for the
underlying WebSocketSession to have access to the same map instance
and it allows the HandshakeHandler to change the attributes even if
it doesn't need to do that today.

Issue: SPR-12716

sbrannen committed:
Preserve ordering of inlined props in @TestPropertySource

The initial implementation for adding inlined properties configured via
@TestPropertySource to the context's environment did not preserve the
order in which the properties were physically declared. This makes
@TestPropertySource a poor testing facility for mimicking the
production environment's configuration if the property source mechanism
used in production preserves ordering of property names -- which is the
case for YAML-based property sources used in Spring Boot, Spring Yarn,
etc.

This commit addresses this issue by ensuring that the ordering of
inlined properties declared via @TestPropertySource is preserved.
Specifically, the original functionality has been refactored. extracted
from AbstractContextLoader, and moved to TestPropertySourceUtils where
it may later be made public for general purpose use in other frameworks.

Issue: SPR-12710
(cherry picked from commit d6a799ad4af81da8840fbd5536efd49b9225ef8a)

sbrannen committed:
Open up TestPropertySourceUtils for public consumption

Spring Framework 4.1 introduced support for @TestPropertySource;
however, the utilities used to parse inlined properties and add test
property sources to the environment are currently private which
prevents reuse by third-party frameworks like Spring Boot.

This commit addresses this issue by making such utilities public.

 - TestPropertySourceUtils is now a public class.

 - Various utility methods in TestPropertySourceUtils have been made
   public.

 - addResourcePropertySourcesToEnvironment() has been renamed to
   addPropertiesFilesToEnvironment().

 - extractEnvironmentProperties() has been renamed to
   convertInlinedPropertiesToMap().

 - All public methods in TestPropertySourceUtils are now fully
   documented.

Issue: SPR-12721
(cherry picked from commit 75e0bc9271e6d5089f6c47f81114d58d09383c7d)

sbrannen committed:
Polish Javadoc for @TestPropertySource

(cherry picked from commit 0267715c9dc50019080039febfd4362894b49aae)

sbrannen committed:
Make TestPropertySourceUtils more robust

 - Added assertions for pre-conditions on method arguments for all
   public utility methods.

 - Introduced additional tests in TestPropertySourceUtilsTests to verify
   the new pre-conditions.

 - Introduced INLINED_PROPERTIES_PROPERTY_SOURCE_NAME constant for the
   name of the MapPropertySource created from inlined properties; the
   name therefore no longer contains the inlined properties, but the
   original values of the inlined properties can now be logged at debug
   level.

 - Simplified tests in InlinedPropertiesTestPropertySourceTests.

Issue: SPR-12721
(cherry picked from commit 42af33034d84327340e4bd53ec3f6c5bba626069)

sdeleuze committed:
Add parametrized constructors to MappingJackson2MessageConverter

Issue: SPR-12724
(cherry picked from commit 8159aa9)

jhoeller committed:
ScheduledAnnotationBeanPostProcessor uses target class as cache key in case of proxy

Issue: SPR-12709
(cherry picked from commit 1273c90)

jhoeller committed:
ServletUriComponentsBuilder.java avoids NPE on scheme check

Issue: SPR-12723
(cherry picked from commit 61cc3b5)

rstoyanchev committed:
Fix handling of empty payload Pong message on Jetty

Issue: SPR-12727

rstoyanchev committed:
Ignore Pong messages in StompSubProtocolHandler

Issue: SPR-12728

jhoeller committed:
Polishing

jhoeller committed:
Latest dependency updates (AspectJ 1.8.5, Joda-Time 2.7, H2 1.4.185)

jhoeller committed:
Latest dependency updates (Groovy 2.3.10, SLF4J 1.7.10)

sbrannen committed:
Polish reference manual regarding WebSocket

 - fixed typos
 - improved grammar and punctuation
 - fixed class names

(cherry picked from commit c7a456c0bfd4edfca7ac7f854c5b0009b659d45c)

jhoeller committed:
MessageHeaderAccessor defensively checks id, timestamp and contentType values

Issue: SPR-12730
(cherry picked from commit dbd353b)

jhoeller committed:
Correct reference to executeAndReturnKey method

Issue: SPR-12639
(cherry picked from commit f20a624)

sdeleuze committed:
Improve error handling in WebUtils.isValidOrigin()

With this commit, WebUtils.isValidOrigin() logs an error message instead
of throwing an IllegalArgumentException when Origin header value is
invalid (for example when it does not contain the scheme).

Issue: SPR-12697

sdeleuze committed:
Adjust log level for invalid SockJS or Websocket requests

jhoeller committed:
Revised common validation methods in AbstractMessageConverterMethodArgumentResolver

The protected validation methods are analogous to ModelAttributeMethodProcessor now.

Issue: SPR-12655
(cherry picked from commit 7191050)

spring-builds committed:
Next Development Version

jhoeller committed:
ScriptUtils properly refers to statement numbers instead of line numbers

Issue: SPR-12752
(cherry picked from commit 8f228d1)

jhoeller committed:
Revised condition override check based on method names instead of bean names

Issue: SPR-12744
(cherry picked from commit bb5b5d5)

jhoeller committed:
Polishing
(cherry picked from commit cc33d3f)

jhoeller committed:
Latest dependency updates (Tomcat 8.0.20, Jetty 9.2.9, EhCache 2.9.1, EhCache-JCache 1.0.1)

jhoeller committed:
Avoid potential deadlocks between event multicaster and singleton registry through shared lock

Issue: SPR-12739
(cherry picked from commit 81102de)

rstoyanchev committed:
Fix white space issues

rstoyanchev committed:
Prevent ISE in the MockMvc PrintingResultHandler

Issue: SPR-12735

bclozel committed:
Allow file locations for resource handling

Prior to this change, location checks for serving resources would append
`/` to the location path it didn't already have one.

This commit makes sure not to append a `/` if the provided location is
actually a file.

Issue: SPR-12747

sbrannen committed:
Refer to statement numbers in ScriptStatementFailedException

ScriptStatementFailedException now properly refers to statement numbers
instead of line numbers.

Issue: SPR-12752
(cherry picked from commit 291b3de72bcf290f07ce901c422e7f5031163e99)

jhoeller committed:
RestTemplate avoids use of warn level for response errors

Protected handleResponse method replaces former private logResponseStatus/handleResponseError methods.

Issue: SPR-12760
(cherry picked from commit 594a14a)

jhoeller committed:
Revised documentation on AUTO_ACKNOWLEDGE behavior

Issue: SPR-12705
(cherry picked from commit c1a9e34)

jhoeller committed:
LoadTimeWeavingConfiguration should not rely on private field injection

Issue: SPR-12776

jhoeller committed:
Polishing
(cherry picked from commit 2b34094)

jhoeller committed:
Polishing
(cherry picked from commit e5207e6)

jhoeller committed:
Fixed javadoc warnings

sbrannen committed:
Include charset in EncodedResource.equals()

Prior to this commit, the implementation of equals() in EncodedResource
was based solely on the resource and encoding. Thus, if a Charset were
specified instead of an encoding, invocations of equals() would not
work as expected.

This commit addresses this issue by including the charset in the
implementation of equals() and introducing corresponding tests in a new
EncodedResourceTests class. Furthermore, this commit makes
EncodedResource immutable and updates all Javadoc to reflect support
for the encoding and charset properties.

Issue: SPR-12767
(cherry picked from commit 93c70b7440783e9b047a52ee3ff9dec5d45300a6)

sbrannen committed:
Support XML config fully in web integration tests

Prior to this commit, it was impossible to use all features of XML
configuration (e.g., the <qualifier> tag) in web-based integration
tests (loaded using @WebAppConfiguration, @ContextConfiguration, etc.)
if the Groovy library was on the classpath. The reason is that the
GroovyBeanDefinitionReader used internally by
GenericGroovyXmlWebContextLoader disables XML validation for its
internal XmlBeanDefinitionReader, and this prevents some XML
configuration features from working properly. For example, the default
value for the 'type' attribute (defined in the spring-beans XSD) of the
<qualifier> tag gets ignored, resulting in an exception when the
application context is loaded.

This commit addresses this issue by refactoring the implementation of
loadBeanDefinitions() in GenericGroovyXmlWebContextLoader to use an
XmlBeanDefinitionReader or GroovyBeanDefinitionReader depending on the
file extension of the resource location from which bean definitions
should be loaded. This aligns the functionality of
GenericGroovyXmlWebContextLoader with the existing functionality of
GenericGroovyXmlContextLoader.

Issue: SPR-12768
(cherry picked from commit 2ba1151b7fa3fcd3b0c964fc377ce57671eabb8b)

sbrannen committed:
Polish Javadoc for GroovyBeanDefinitionReader

sbrannen committed:
Fully support XML config in GroovyBeanDefinitionReader

Prior to this commit, the GroovyBeanDefinitionReader claimed (via its
Javadoc) that it fully supported XML configuration files in addition to
its Groovy DSL; however, this was unfortunately inaccurate since XML
validation was disabled by default which led to certain features of XML
configuration not working. For example, it was impossible to define a
<qualifier> in an XML config file without specifying the 'type'
attribute (which has a default value defined in the spring-beans XSD).

This commit fixes this issue by ensuring that bean definitions in XML
resources are loaded with a "standard" XmlBeanDefinitionReader that is
created with default settings (i.e., with XML validation enabled). With
regard to backwards compatibility, bean definitions defined using the
Groovy DSL are still loaded with an XmlBeanDefinitionReader that has
XML validation disabled by default which is necessary for proper
parsing of the Groovy DSL.

Issue: SPR-12769
(cherry picked from commit 7edc7c2c8f3e98a9c61d4c9520489d132c00e0a0)

jhoeller committed:
Fixed javadoc

jhoeller committed:
Latest dependency updates (SnakeYAML 1.15, FreeMarker 2.3.22, JRuby 1.7.19, JAMon 2.81, JiBX 1.2.6, XMLUnit 1.6)

jhoeller committed:
Latest dependency updates (XMLUnit 1.6, JsonPath 1.2, H2 1.4.186)

jhoeller committed:
Converter interface explicitly documents null values

Issue: SPR-12780
(cherry picked from commit dcb1145)

jhoeller committed:
ObjectToOptionalConverter preserves existing Optional instances

Issue: SPR-12785
(cherry picked from commit f786fc3)

rstoyanchev committed:
UriComponentsBuilder.fromHttpRequest sets port correctly

Cherry-picked from commit d64c48 on master

Issue: SPR-12771

jhoeller committed:
getTypeForFactoryBean suppresses instantiation failures for non-singleton FactoryBeans

Issue: SPR-12786
(cherry picked from commit 9b25d6a)

jhoeller committed:
Polishing
(cherry picked from commit b541fc9)

jhoeller committed:
Dependency updates for spring-oxm generation tasks
(cherry picked from commit f8fd19d)

jhoeller committed:
Polishing

sbrannen committed:
Support @NumberFormat as a meta-annotation

This commit ensures that @NumberFormat can be used as a
meta-annotation, as was already the case for @DateTimeFormat.

In addition, this commit polishes FormattingConversionServiceTests and
MvcNamespaceTests.

Issue: SPR-12743
(backported from commits df1d90f & c3408d8)

snicoll committed:
Call AsyncUncaughtExceptionHandler when necessary

If a sub-class of Future (such as ListenableFuture) is used as a return
type and an exception is thrown, the AsyncUncaughtExceptionHandler is
called. Now checking for any Future implementation instead of a faulty
strict matching.

Issue: SPR-12797

jhoeller committed:
PropertyResourceConfigurerTests accepts "." in operating system name

Issue: SPR-12794
(cherry picked from commit 6fe85c2)

jhoeller committed:
StandardMultipartFile declares itself as Serializable now

Issue: SPR-12795
(cherry picked from commit 5ba7b89)

jhoeller committed:
ResponseEntity's headers(HttpHeaders) accepts null value

Issue: SPR-12792
(cherry picked from commit 73e8021)

jhoeller committed:
Fine-tuned varargs handling in ReflectiveMethodResolver (including useDistance mode)

Issue: SPR-12803
Issue: SPR-12808

jhoeller committed:
Explicit note on change of read-only handling in 4.1

Issue: SPR-8959

bclozel committed:
Fix InputStream caching in ContentCachingReqWrapper

Prior to this commit, the ContentCachingRequestWrapper would immediately
consume the wrapped request's InputStream when asked for the cached
content; that caused several issues:

* the request body was read in memory even if it wasn't yet consumed by
the application, leading to inefficiencies.
* when requesting the InputStream, an empty InputStream was returned
since the original was already read.

This case only happened for form POSTs requests.

This commit makes sure that the wrapper does not alter the request
expected behavior:

* when getting the inputstream, it is wrapped in order to cache its
content
* when getting request parameters, the request body is cached and its
inputstream is consumed, as expected

Issue: SPR-12810

rstoyanchev committed:
Move X-Forwarded-* tests to UriComponentsBuilderTests

rstoyanchev committed:
Support comma-separated X-Forwarded-Port

Issue: SPR-12813

rstoyanchev committed:
Support comma-separated X-Forwarded-Proto

Issue: SPR-12816

jhoeller committed:
Polishing (in particular updating javadoc references to Apache Commons)
(cherry picked from commit bc6a98c)

jhoeller committed:
ContentCachingRequestWrapper converts IOException to IllegalStateException

Issue: SPR-12810

jhoeller committed:
Explicit documentation on MutablePersistenceUnitInfo's addManagedPackage

Issue: SPR-12821
(cherry picked from commit 4f1d9fd)

jhoeller committed:
Explicit documentation on 4.1+ properties in AbstractMessageListenerContainer
(cherry picked from commit 6f98cf3)

jhoeller committed:
Revised DefaultManagedAwareThreadFactory to make its non-JNDI fallback work

Issue: SPR-12830
(cherry picked from commit a3e5fbf)

jhoeller committed:
Optimized access to resolved bean type (avoiding BeanFactory locks)

Revised HandlerMethod.getBeanType() impl for both web and messaging.
In addition, HandlerMethods get created with the internal BeanFactory now.

Issue: SPR-12832
(cherry picked from commit 898c24f)

jhoeller committed:
BeansDtdResolver resolves spring-beans.dtd declarations to spring-beans-2.0.dtd file

Issue: SPR-12836
(cherry picked from commit 8e4bfa9)

jhoeller committed:
Polishing
(cherry picked from commit 162ee36)

jhoeller committed:
Latest dependency updates (Groovy 2.3.11, Jetty 9.2.10, Netty 4.0.26, Undertow 1.1.3)

snicoll committed:
Restore proper use of CacheLoader

Since Guava 11, CacheLoader is only invoked with a LoadingCache but the
GuavaCache wrapper is always invoking getIfPresent(), available on the
main Guava Cache interface.

Update GuavaCache#get to check for the presence of a LoadingCache and
call the appropriate method.

Issue: SPR-12842
(cherry picked from commit 9172a6d)

jhoeller committed:
Deferred import processing reliably detects late registration attempts

Issue: SPR-12838
(cherry picked from commit 1a8c6fa)

jhoeller committed:
Polishing

jhoeller committed:
Alignment with ResolvableType revision in master

jhoeller committed:
Removed outdated references to Servlet 2.3/2.4

jhoeller committed:
TcpConnection extends java.io.Closeable

Issue: SPR-12849
(cherry picked from commit 4e28da4)

jhoeller committed:
Fixed addConverterFactory assertion
(cherry picked from commit 22670b7)

jhoeller committed:
Consistent javadoc param declarations for type variables
(cherry picked from commit d23893f)

snicoll committed:
Only require an exception CacheResolver if necessary

Previously, a cache infrastructure with only a CacheResolver would have
worked fine until the JSR-107 API is added to the classpath. When this is
the case, the JCache support kicks in and an exception cache resolver is
all of the sudden required.

The CacheResolver _is_ different as the default implementation does look
different attributes so if a custom CacheResolver is set, it is not
possible to "reuse" it as a fallback exception CacheResolver.

Now, an exception CacheResolver is only required if a JSR-107 annotation
with an "exceptionCacheName" attribute is processed (i.e. the exception
CacheResolver is lazily instantiated if necessary).

The use case of having a CachingConfigurerSupport with only a
CacheResolver was still broken though since the JCache support only looks
for a JCacheConfigurer bean (per the generic type set on
AbstractCachingConfiguration). This has been fixed as well.

Issue: SPR-12850
(cherry picked from commit 314b069)

jhoeller committed:
Polishing

spring-builds committed:
Next Development Version

jhoeller committed:
Clarification: Lifecycle does not imply auto-startup semantics

Issue: SPR-12855
(cherry picked from commit 7e22623)

jhoeller committed:
BeanWrapperImpl.setPropertyValue throws InvalidPropertyException with correct property value for nested collection case

Issue: SPR-12866
(cherry picked from commit e377fc0)

jhoeller committed:
AnnotationUtils.getAnnotationAttributes makes Method accessible (analogous to AnnotationUtils.getValue)

Issue: SPR-12858
(cherry picked from commit 1d33fd0)

jhoeller committed:
Latest dependency updates (Reactor 1.1.6, SLF4J 1.7.11, Tomcat 8.0.21)

jhoeller committed:
Polishing

poutsma committed:
Netty4ClientHttpRequest ignores query params

Before this commit, Netty4ClientHttpRequest ignored query parameters
(?foo=bar). This commit fixes that.

Issue: SPR-12779
(cherry picked from commit caee78a)

bclozel committed:
Fix client-library-url ignored in MVC namespace

Prior to this commit, the `client-library-url` XML attribute was not
effective in the MVC namespace, leaving the default value configured:

```xml
<websocket:sockjs client-library-url="/js/sockjs.js" />
```

This commit fixes the sockjs namespace handler and makes sure that this
attribute is configured on the `SockJsService` Bean to be created.

Issue: SPR-12874

jhoeller committed:
CronSequenceGenerator explicitly rejects invalid incrementer delta

Issue: SPR-12871
(cherry picked from commit ceb17fc)

jhoeller committed:
Polishing

jhoeller committed:
AnnotatedElementUtils leniently ignores TypeNotPresentExceptions (just like AnnotationUtils)

Also refines logIntrospectionFailure to just log at debug level for meta-annotation introspection failures.

Issue: SPR-12889
(cherry picked from commit e78b086)

jhoeller committed:
AnnotationAsyncExecutionAspect properly accepts ListenableFuture return type

Issue: SPR-12895
(cherry picked from commit 8b2d995)

jhoeller committed:
HttpEntity and its subclasses insist on same target type for equality

Issue: SPR-12910
(cherry picked from commit af272c2)

jhoeller committed:
@Bean-returned FactoryBean proxy delegates to actual target instance now

Issue: SPR-12915
(cherry picked from commit 1da98b0)

jhoeller committed:
FactoryBean type check logs currently-in-creation exception at debug level

Issue: SPR-12900
(cherry picked from commit 65ba72f)

jhoeller committed:
InvocableHandlerMethod logs argument resolution exceptions at debug level

Issue: SPR-12925
(cherry picked from commit 10a51a4)

jhoeller committed:
StatementCreatorUtils.setValue only uses setString etc for Types.OTHER in case of Oracle

Issue: SPR-12890
(cherry picked from commit 050e581)

snicoll committed:
Apply DestinationResolver to listener container

Previously, a custom `DestinationResolver` was not made available
to the underlying `MessageListener`: if a reply needs to be sent, the
default `DestinationResolver` was used.

This commit ensures that if a custom `DestinationResolver` has been set,
it is also configured accordingly for replies.

Issue: SPR-12927

jhoeller committed:
PathMatchingResourcePatternResolver skips invalid jar classpath entries

Issue: SPR-12928
(cherry picked from commit 49f3046)

bclozel committed:
Fix CachingResourceResolver key generation

When used in combination with GzipResourceResolver, the
CachingResourceResolver does not properly cache results, since it only
takes the request path as a input for cache key generation.

Here's an example of that behavior:

1. an HTTP client requests a resource with `Accept-Encoding: gzip`, so
the GzipResourceResolver can resolve a gzipped resource.
2. the configured CachingResourceResolver caches that resource.
3. another HTTP client requests the same resource, but it does not
support gzip encoding; the previously cached gzipped resource is still
returned.

This commit uses the presence/absence of gzip encoding support as an
input in cache keys generation.

Issue: SPR-12982

rstoyanchev committed:
Allow "ws" and "wss" for isValidCorsOrigin checks

Issue: SPR-12956

sdeleuze committed:
Fix JSON Views code sample in the documentation

Issue: SPR-12994

jhoeller committed:
Fix for ASM issue 317555

Issue: SPR-12957
(cherry picked from commit 2c043ec)

jhoeller committed:
Avoid infinite loop in PatternMatchUtils

Issue: SPR-12971
(cherry picked from commit db80378)

sbrannen committed:
Remove duplicate separators when combining paths

Prior to this commit, AntPathMatcher would not correctly combine a path
that ends with a separator with a path that starts with a separator.
For example, `/foo/` + `/bar` combined into `/foo//bar`.

Specifically, this commit:

 - Removes the duplicated separator in combined paths

 - Improves RequestMappingInfo's toString() representation

 - Fixes Javadoc formatting in AntPathMatcher

 - Polishes AntPathMatcherTests

 - Polishes Javadoc in AbstractRequestCondition

Issue: SPR-12975
(cherry picked from commit a80d0e87a481242bc0a32aae26e0135f98913596)

jhoeller committed:
CustomBooleanEditor avoids potential NPE in case of allowEmpty=false

Issue: SPR-13010
(cherry picked from commit 549e888)

jhoeller committed:
Polishing

jhoeller committed:
Latest dependency updates (Hibernate 4.3.9, Jackson 2.4.6, JasperReports 6.0.4, Netty 4.0.27, SLF4J 1.7.12, Tomcat 8.0.22, Undertow 1.1.4)

jhoeller committed:
Javadoc fixes

jhoeller committed:
RequestResponseBodyMethodProcessor decorates ServletInputStream as well

Issue: SPR-13014

jhoeller committed:
Polishing

jhoeller committed:
SimpleAliasRegistry prevents NPE for alias resolved to null

Issue: SPR-13016
(cherry picked from commit e83d495)

bclozel committed:
Fix empty URLs handling in ResourceUrlEncodingFilter

Prior to this commit, the ResourceUrlEncodingFilter would fail with a
StringIndexOutOfBoundsException when:

* the current request has a servlet context
* the URL to encode is relative and is shorter than the context value

This change defensively checks for those lengths and delegates to the
parent implementation if necessary.

Issue: SPR-13018

jhoeller committed:
Actual hasText assertion in SockJsFrame

Issue: SPR-13019
(cherry picked from commit 02d28ae)

jhoeller committed:
PropertyOrFieldReference avoids NPE through defensive copy of volatile accessor reference

Issue: SPR-13023
(cherry picked from commit 9799df3)

jhoeller committed:
Polishing
(cherry picked from commit 0711d6d)

rstoyanchev committed:
Fix typo in reference

Issue: SPR-13043

bclozel committed:
Fix directories I/O in ResourceHttpRequestHandler

Prior to this commit, the `ResourceHttpRequestHandler` would not
properly handle HTTP requests to **directories contained in JARs**.
This would result in HTTP 500 errors, caused by `FileNotFoundException`
or `NullPointerException`.

This can be tracked to webapp ClassLoader implementations in servlet
containers:
* in Jetty9x, fetching a directory within a JAR as a `Resource` and
getting its InputStream work fine, but attempting to `close()` it
results in a NullPointerException as the underlying stream is null.
* In Tomcat6x, one cannot fetch an InputStream for the same `Resource`
as it throws a FileNotFoundException.

This change adds more try/catch clauses and catches more Exception so as
to result in HTTP 200 OK responses instead of server errors. While this
is inconsistent because the same code path would result in HTTP 404 with
existing directories on the file system, there's no other simple way to
make those checks for resources contained in JARs.

Issue: SPR-12999

snicoll committed:
Add possible matches for field access

DirectFieldAccessor now support richer NotWritablePropertyException
content, including dedicated error message and possible matches.

Issue: SPR-13053

bclozel committed:
Fix SPR-12999 backport (missing test dependency)

snicoll committed:
Make PropertyMatches public

Issue: SPR-13054

jhoeller committed:
ByteBufferConverter explicitly declares applicability to byte[]

Includes an optimization for simple ByteBuffer duplication.

Issue: SPR-13056
(cherry picked from commit 792b7b9)

jhoeller committed:
Message listener containers only call Session.recover() in client acknowledge mode

Issue: SPR-13052
(cherry picked from commit fee63fd)

jhoeller committed:
Polishing

jhoeller committed:
Latest dependency updates (Hibernate ORM 4.3.10, Netty 4.0.28)

jhoeller committed:
Polishing

jhoeller committed:
SockJsTransportFailureException provides constructor variant without session id
(cherry picked from commit e0a11f2)

jhoeller committed:
Polishing
(cherry picked from commit 92bf32b)

jhoeller committed:
SpringSessionSynchronization's beforeCompletion unbinds after disconnect failure as well

Issue: SPR-13089
(cherry picked from commit 8c337a4)

jhoeller committed:
Lite configuration candidate check defensively handles method introspection failure

Issue: SPR-13091
(cherry picked from commit 4f1286a)

jhoeller committed:
Latest dependency updates (Jetty 9.2.11, Tomcat 8.0.23)

snicoll committed:
Allow Cache annotations to not specify any cache name

Since Spring 4.1, a CacheResolver may be configured to customize the way
the cache(s) to use for a given cache operation are retrieved. Since a
CacheResolver implementation may not use the cache names information at
all, this attribute has been made optional.

However, a fix was still applied, preventing a Cache operation without a
cache name to be defined properly. We now allow this valid use case.

Issue: SPR-13081

snicoll committed:
Polish

Update copyright headers

bclozel commented:
I don't see any commit of yours in this PR.
Could you fix this PR please? New features should be committed against master directly - backporting to 4.1.x is a decision for the team to make.

Thanks!

sbrannen committed:
Proper support for Root WAC in Spring MVC Test

The modifications to DefaultMockMvcBuilder performed in conjunction
with SPR-12553 introduced a breaking change: the WebApplicationContext
supplied to DefaultMockMvcBuilder's constructor was *always* stored in
the ServletContext as the root WebApplicationContext, overwriting a
root WebApplicationContext that had been set by the user or by the
Spring TestContext Framework (TCF) -- for example, in
AbstractGenericWebContextLoader. Consequently, the changes in SPR-12553
cause tests that use @ContextHierarchy to fail if web components rely
on the correct WebApplicationContext being stored under the
WebApplicationContext#ROOT_WEB_APPLICATION_CONTEXT_ATTRIBUTE key.

This commit reverts the breaking changes introduced in SPR-12553: if
the root WebApplicationContext has already been set in the
ServletContext of the WebApplicationContext supplied to
DefaultMockMvcBuilder, no action is taken.

Furthermore, this commit introduces new code to address the initial
intent of SPR-12553. Specifically, if the root WebApplicationContext
has NOT been set in the ServletContext of the WebApplicationContext
supplied to DefaultMockMvcBuilder, the application context hierarchy
will be traversed in search of the root WebApplicationContext, and the
root WebApplicationContext will then be stored under the
ROOT_WEB_APPLICATION_CONTEXT_ATTRIBUTE key.

Issue: SPR-13075, SPR-12553
(cherry picked from commit f6d2fe471a26fb6ff5894480f50dd55365e62f06)

jhoeller committed:
Reverted enhanceFactoryBean revision in 4.1.x (making it 4.2 only)

Issue: SPR-12915
Issue: SPR-13095

jhoeller committed:
Latest dependency updates (AspectJ 1.8.6, Joda-Time 2.8)

jhoeller committed:
Enabled test for AspectJ pointcut matching against lambda-defined bean

Issue: SPR-11807
(cherry picked from commit b02352d)

jhoeller committed:
AspectJExpressionPointcut defensively catches exceptions thrown from ShadowMatch.matchesJoinPoint

Issue: SPR-13102
(cherry picked from commit 75edb39)

jhoeller committed:
WebAsyncUtils avoids reflection for creating StandardServletAsyncWebRequest

Issue: SPR-13112
(cherry picked from commit 32b9ea9)

jhoeller committed:
Stronger warning about lookup methods not working with @Bean

Includes an updated CGLIB AOP proxy note on constructor invocations.

Issue: SPR-13108
Issue: SPR-13103

jhoeller committed:
Polishing

jhoeller committed:
ConfigurationClassParser ignores unresolvable member classes

Issue: SPR-13115
(cherry picked from commit 1c01f57)

jhoeller committed:
Latest dependency updates (H2 1.4.187, Joda-Time 2.8.1)

jhoeller committed:
Polishing

jhoeller committed:
TransactionAttributeSourcePointcut pointcut skips target classes with TransactionalProxy marker (e.g. Spring Data proxies)

Issue: SPR-13109

jhoeller committed:
Fixed fallback mode in ObjenesisCglibAopProxy

Issue: SPR-13131

jhoeller committed:
Polishing

jhoeller committed:
ObjenesisCglibAopProxy catches NoClassDefFoundError as well

Issue: SPR-13131

martinlippert committed:
Fix exception in AntPathMatcher for leading *

Issue: SPR-13139

bclozel commented:
Closing this PR as invalid.
Currently emailing with the PR author to create a proper one.
